CodeJam.Main/Assertions/DebugCode.generated.cs

```diff
@@ -149,15 +155,46 @@ public static void NotEmpty<T>([NotNull] T[] arg, [NotNull, InvokerParameterName
 		/// <param name="argName">Name of the argument.</param>
 		[Conditional(DebugCondition), DebuggerHidden, MethodImpl(AggressiveInlining)]
 		[AssertionMethod]
-		[ContractAnnotation("arg:null => stop")]
+		[ContractAnnotation("arg: null => halt")]
 		public static void NotNullNorWhiteSpace(
-			[CanBeNull, AssertionCondition(AssertionConditionType.IS_NOT_NULL)] string arg,
+			[CanBeNull] string arg,
 			[NotNull, InvokerParameterName] string argName)
 		{
 			if (arg.IsNullOrWhiteSpace())
 				throw CodeExceptions.ArgumentNullOrWhiteSpace(argName);
 		}
 
+		/// <summary>Ensures that <paramref name="arg"/> and its all items != <c>null</c></summary>
+		/// <typeparam name="T">Type of the value. Auto-inferred in most cases</typeparam>
+		/// <param name="arg">The argument.</param>
+		/// <param name="argName">Name of the argument.</param>
+		[Conditional(DebugCondition), DebuggerHidden, MethodImpl(AggressiveInlining)]
+		[AssertionMethod]
+		[ContractAnnotation("arg: null => halt")]
+		public static void NotNullAndItemNotNull<T>(
+			[CanBeNull, InstantHandle] IEnumerable<T> arg,
+			[NotNull, InvokerParameterName] string argName) where T : class
+		{
+			if (arg == null)
+				throw CodeExceptions.ArgumentNull(argName);
+			ItemNotNull(arg, argName);
+		}
+
+		/// <summary>Ensures that all items in <paramref name="arg"/> != <c>null</c></summary>
+		/// <typeparam name="T">Type of the value. Auto-inferred in most cases</typeparam>
+		/// <param name="arg">The argument.</param>
+		/// <param name="argName">Name of the argument.</param>
+		[Conditional(DebugCondition), DebuggerHidden, MethodImpl(AggressiveInlining)]
+		[AssertionMethod]
+		public static void ItemNotNull<T>(
+			[NotNull, InstantHandle] IEnumerable<T> arg,
```

Comment:
What you think about CollectionAccessType﻿.attribute?

Reply:
As far as I know that's for collection implementation, not for external methods.
> Provides a value for the CollectionAccessAttribute to define how **the collection method invocation** affects the contents of the collection.

[(c)](https://www.jetbrains.com/help/resharper/Reference__Code_Annotation_Attributes.html#CollectionAccessType)